### PR TITLE
add a bind-match-failure warning to the window widget's hit method

### PR DESCRIPTION
**Pull request by gleachkr**
Here's a first pass at closing #392

### CODE REVIEW

File: lib/window.lua [hunk 1]
```diff
@@ -262,7 +262,14 @@ _M.methods = {
 
         local caught, newbuf = lousy.bind.hit(w, w.binds, mods, key, opts)
         if w.win then -- Check binding didn't cause window to exit
+            local oldbuf
+            if w.buffer then oldbuf = w.buffer .. key else oldbuf = key end
```
Comment (aidanholm) [inline lib/window.lua]: Should probably be:
```lua
local oldbuf = (w.buffer or "") .. key
```

File: lib/window.lua [hunk 2]
```diff
@@ -262,7 +262,14 @@ _M.methods = {
...
             w.buffer = newbuf
+            if not (newbuf or caught) and key:len() == 1 and w:is_mode("normal") then
```
Comment (aidanholm) [inline lib/window.lua]: What happens without the `w:is_mode("normal")` check? My concern is for things like `ex-follow` mode; pressing `;` followed by some random key wouldn't show anything in this case.

File: lib/window.lua [hunk 3]
```diff
@@ -262,7 +262,14 @@ _M.methods = {
...
+                w:notify(oldbuf .. " is not a prefix of an existing binding")
```
Comment (aidanholm) [inline lib/window.lua]: This can probably just be "No binding matched..."

**Review by aidanholm — CHANGES_REQUESTED** [review]
That's a lot less code than I was expecting!

File: lib/window.lua (hunk 2 above)
Comment (gleachkr) [reply lib/window.lua]: It tends to spam the user with no-binding notifications in the non-normal modes, like insert or command. Maybe this should be configurable?

File: lib/window.lua (hunk 3 above)
Comment (gleachkr) [reply lib/window.lua]: How about `"No binding matched by " .. oldbuf`, so that you can tell if it's a typo?

File: lib/window.lua (hunk 2 above)
Comment (aidanholm) [reply lib/window.lua]: You could try moving it to the signal handler around line 169 of window.lua

File: lib/window.lua (hunk 3 above)
Comment (aidanholm) [reply lib/window.lua]: Good thinking, but s/by/for/

File: lib/window.lua (hunk 2 above)
Comment (gleachkr) [reply lib/window.lua]: Same problem there, I think, though I might not be seeing the idea you're seeing.

Also, I'm not sure in that context how to tell when a key is part of a prefix for a binding, and when it's not---it looks like you get `match` false when the keypress is part of a prefix, and just hasn't matched yet.

**Comment by gleachkr** [comment]
Went ahead and applied those suggestions, also adding an option to make the modes in which the warning is triggered extensible (with default value `{"normal","ex-follow"}`).

**Comment by aidanholm** [comment]
I actually don't think that's a good idea; ideally, it'd just work in _all_ modes. I'd say keep the whitelist private, which means no API break if this gets improved later.

**Comment by aidanholm** [comment]
Instead of checking the current mode, you could check whether the input bar or the current view is focused; it's hacky, but it should work.

**Comment by gleachkr** [comment]
Good idea. Still need a special case for insert mode, but seems to work otherwise. Added to latest commit.

File: lib/window.lua [hunk 4]
```diff
@@ -262,7 +262,14 @@ _M.methods = {
 
         local caught, newbuf = lousy.bind.hit(w, w.binds, mods, key, opts)
         if w.win then -- Check binding didn't cause window to exit
+            local oldbuf = (w.buffer or "") .. key
             w.buffer = newbuf
+
```
Comment (aidanholm) [inline lib/window.lua]: remove whitespace in this line

File: lib/window.lua [hunk 5]
```diff
@@ -262,7 +262,14 @@ _M.methods = {
 
         local caught, newbuf = lousy.bind.hit(w, w.binds, mods, key, opts)
         if w.win then -- Check binding didn't cause window to exit
+            local oldbuf = (w.buffer or "") .. key
             w.buffer = newbuf
+            
+            if not (newbuf or caught or w.ibar.input.focused or w:is_mode("insert")) and key:len() == 1 then
```
Comment (aidanholm) [inline lib/window.lua]: key can be a number, for mouse button bindings:
```
[    6.764182] E [core/common/lualib]: Lua error: attempt to index local 'key' (a number value)
                 Traceback:
                 (1) lib/window.lua:268 in function hit
                 (2) lib/webview.lua:82 in function [anonymous]
```

**Review by aidanholm — CHANGES_REQUESTED** [review]

File: lib/window.lua [hunk 6]
```diff
@@ -262,7 +262,14 @@ _M.methods = {
...
+                w:notify("no binding matched for " .. oldbuf)
+            else
+                w:set_prompt()
```
Comment (aidanholm) [inline lib/window.lua]: With this, any notification will disappear if the user scrolls with the mousewheel, but clicks are ignored
I think scrolling with the mousewheel shouldn't dismiss the notice, and clicks should

Comment (gleachkr) [reply lib/window.lua]: Yes, that seems reasonable. I don't have a scrollwheel right now. What keys does a scrollwheel send? or is there a better way of detecting scrollwheel events?

File: lib/window.lua (hunk 5 above)
Comment (gleachkr) [reply lib/window.lua]: Right. Should there be a warning when no mouse button binding is matched (perhaps unless it's just a vanilla click), or should mouse events be ignored?

File: lib/window.lua (hunk 6 above)
Comment (aidanholm) [reply lib/window.lua]: If you have a touchpad, that works the same way. Scrolling is done with synthesized key events with `key = "Scroll"`: see `lib/webview.lua:88`

File: lib/window.lua (hunk 5 above)
Comment (aidanholm) [reply lib/window.lua]: I think ignoring clicks is fine

**Comment by gleachkr** [comment]
OK, updated.

File: lib/window.lua [hunk 7]
```diff
@@ -262,7 +262,14 @@ _M.methods = {
 
         local caught, newbuf = lousy.bind.hit(w, w.binds, mods, key, opts)
         if w.win then -- Check binding didn't cause window to exit
+            local oldbuf = (w.buffer or "") .. key
             w.buffer = newbuf
+            if not (newbuf or caught or w.ibar.input.focused or w:is_mode("insert"))
+                and type(key) == "string" and key:len() == 1 then
+                w:notify("no binding matched for " .. oldbuf)
+            elseif not (key == "Scroll") then
```
Comment (aidanholm) [inline lib/window.lua]: `key ~= "Scroll"`

**Review by aidanholm — CHANGES_REQUESTED** [review]
final code nit :)

**Comment by gleachkr** [comment]
Great! Updated along with changelog.

**Comment by aidanholm** [comment]
Merged :tada: Thanks :)

**Pull request merged**